pkg/auth/handler/sso/auth_url.go

```diff
@@ -117,51 +122,125 @@ func (p AuthURLRequestPayload) Validate() error {
-	return true
+func (h *AuthURLHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	result, err := handler.Transactional(h.TxContext, func() (interface{}, error) {
+		return h.Handle(w, r)
+	})
+	if err != nil {
+		handler.WriteResponse(w, handler.APIResponse{Err: skyerr.MakeError(err)})
+		return
+	}
+	if r.Method == http.MethodPost {
+		handler.WriteResponse(w, handler.APIResponse{Result: result})
+		return
+	}
+	http.Redirect(w, r, result.(string), http.StatusFound)
 }
 
-func (h AuthURLHandler) DecodeRequest(request *http.Request) (handler.RequestPayload, error) {
+func (h *AuthURLHandler) Handle(w http.ResponseWriter, r *http.Request) (result interface{}, err error) {
 	payload := AuthURLRequestPayload{
 		// avoid nil pointer
-		Scope:   make([]string, 0),
 		Options: make(sso.Options),
 	}
-	err := json.NewDecoder(request.Body).Decode(&payload)
-	payload.UXMode = sso.UXModeFromString(payload.RawUXMode)
+	if r.Method == http.MethodPost {
+		err = json.NewDecoder(r.Body).Decode(&payload)
+		if err != nil {
+			return
+		}
+	} else {
+		err = r.ParseForm()
+		if err != nil {
+			return
+		}
+		payload.CallbackURL = r.Form.Get("callback_url")
+		payload.UXMode = sso.UXMode(r.Form.Get("ux_mode"))
+		payload.MergeRealm = r.Form.Get("merge_realm")
+		payload.OnUserDuplicate = sso.OnUserDuplicate(r.Form.Get("on_user_duplicate"))
+	}
 
-	return payload, err
-}
+	if payload.MergeRealm == "" {
+		payload.MergeRealm = password.DefaultRealm
+	}
+
+	if payload.OnUserDuplicate == "" {
+		payload.OnUserDuplicate = sso.OnUserDuplicateDefault
+	}
+
+	err = payload.Validate()
+	if err != nil {
+		return
+	}
 
-func (h AuthURLHandler) Handle(req interface{}) (resp interface{}, err error) {
 	if h.Provider == nil {
-		err = skyerr.NewInvalidArgument("Provider is not supported", []string{h.ProviderName})
+		err = skyerr.NewInvalidArgument("Provider is not supported", []string{h.ProviderID})
+		return
+	}
+
+	if !h.PasswordAuthProvider.IsRealmValid(payload.MergeRealm) {
+		err = skyerr.NewInvalidArgument("Invalid MergeRealm", []string{payload.MergeRealm})
+		return
+	}
+
+	if !sso.IsAllowedOnUserDuplicate(
+		h.OAuthConfiguration.OnUserDuplicateAllowMerge,
+		h.OAuthConfiguration.OnUserDuplicateAllowCreate,
+		payload.OnUserDuplicate,
+	) {
+		err = skyerr.NewInvalidArgument("Disallowed OnUserDuplicate", []string{string(payload.OnUserDuplicate)})
+		return
+	}
+
+	if e := sso.ValidateCallbackURL(h.OAuthConfiguration.AllowedCallbackURLs, payload.CallbackURL); e != nil {
+		err = skyerr.NewInvalidArgument(e.Error(), []string{string(payload.CallbackURL)})
 		return
 	}
-	payload := req.(AuthURLRequestPayload)
+
+	cookie, cookieErr := r.Cookie(coreHttp.CookieNameOpenIDConnectNonce)
+	if cookieErr == http.ErrNoCookie {
+		// Set nonce
+		nonce := sso.GenerateOpenIDConnectNonce()
+		cookie = &http.Cookie{
+			Name:     coreHttp.CookieNameOpenIDConnectNonce,
+			Value:    nonce,
+			Path:     "/",
+			HttpOnly: true,
+		}
+		http.SetCookie(w, cookie)
```

Comment:
So we expect all SDK would have a cookie jar implemented to use OIDC providers?

Reply:
A simple memory cookie jar is enough for this flow. Cookie is part of the HTTP spec and it does not require any special support from our SDK.